### PR TITLE
Blog app: Validations and spec

### PR DESCRIPTION
**Pull request by Marcraphael12**
### Project requirements
- Add the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Add unit specs for all of your models' methods and validations.

### CODE REVIEW

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @Marcraphael12, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Passing Test Cases :heavy_check_mark: 
Added Validations To Models As Requested Here #3 :heavy_check_mark: 
Descriptive Pr :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._
- [ ] Please do not open a new PR, make required changes here and resubmit for another review
## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb [hunk 1]
```diff
@@ -11,35 +11,38 @@
...
     expect(subject).to_not be_valid
   end
 
   it 'nil should be allowed' do
     author = User.new(Name: 'Tom', Photo: 'https://unsplash.com/photos/F_-0BxGuVvo', Bio: 'Teacher from Mexico.')
     subject = Post.new(author:, Title: 'Hello', Text: 'This is my first post')
-    subject.likes_counter = nil
+    subject.LikesCounter = nil
     expect(subject).to be_valid
   end
```
[the remark this answers is not in the document]
Comment (Marcraphael12) [reply spec/models/post_spec.rb:33]: done

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -5,25 +5,25 @@
```
[the remark this answers is not in the document]
Comment (Marcraphael12) [reply spec/models/user_spec.rb:5]: already have a method to check thath

[the remark this answers is not in the document]
Comment (Marcraphael12) [reply spec/models/user_spec.rb:5]: in post

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -5,25 +5,25 @@
...
-    subject.posts_counter = -1
+    subject.PostCounter = -1
     expect(subject).to_not be_valid
   end
 
   it 'nil should be allowed' do
-    subject.posts_counter = nil
+    subject.PostCounter = nil
     expect(subject).to be_valid
   end
```
[the remark this answers is not in the document]
Comment (Marcraphael12) [reply spec/models/user_spec.rb]: done

File: app/models/post.rb [hunk 4]
```diff
@@ -6,11 +6,21 @@ class Post < ApplicationRecord
...
   # model and the `Comment` and `Like` models, respectively.
-  has_many :comments
-  has_many :likes
+  has_many :comments, foreign_key: 'post_id'
+  has_many :likes, foreign_key: 'post_id'
+
+  # These lines of code are defining validations for the `Post` model.
+  validates :Title, presence: true, length: { maximum: 250 }
+  validates :LikesCounter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }, allow_nil: true
+  validates :CommentsCounter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }, allow_nil: true
```
Comment (Gambit142) [inline app/models/post.rb:16]: - [ ] The previous reviewer requested that it is not a good practice to allow nil values for the `LikesCounter` and the `CommentsCounter`. Allowing `nil` value is against the project's requirement as these counters should be of `integer` data types and `nil` is a different data type. In a real-life setting, the Postcounter will be from number zero and upwards. Kindly implement this change. 👍 😎

File: app/models/user.rb [hunk 5]
```diff
@@ -5,6 +5,12 @@ class User < ApplicationRecord
   has_many :comments, foreign_key: :author_id
   has_many :likes, foreign_key: :author_id
 
+  # validating that the `Name` attribute of the `User`
+  # model is present, meaning it cannot be blank or nil.
+  validates :Name, presence: true
+
+  validates :PostCounter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }, allow_nil: true
+
```
Comment (Gambit142) [inline app/models/user.rb:13]: - [ ] The previous reviewer requested that it is not a good practice to allow nil values for the `PostCounter` Allowing `nil` value is against the project's requirement as this counter should be of `integer` data type and `nil` is a different data type. Kindly implement this change. In a real-life setting, the Postcounter will be from number zero and upwards. 👍 😎

File: spec/models/user_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gambit142) [reply spec/models/user_spec.rb:5]: - [ ] It is good that you are testing it in another file but based on the requirement, all methods in a class should be tested in their `spec` file. This also improves the organization and each class and its methods can be tested in a `spec` file. Kindly implement the change. 👍

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @Marcraphael12 👋, 

Great work on making the changes requested by a previous reviewer 👏 

## Highlights
- Setup linters correctly ✔️ 
- Use of Rspec to test models ✔️ 
- Wrote validations for model attributes ✔️ 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Munsa1) [reply spec/models/user_spec.rb:5]: - [ ] No in the user model you have a `update_posts_counter` method that has not been tested in the `user_spec.rb` file.

Your user.rb file
image

Your user_spec file
image

**Review by Reem-lab — APPROVED** [review]
# Hi @Marcraphael12, 👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  Followed Gitflow 💯
- [ ]  Descrtipive PR⭐
- [ ] Kudos for working on the validations🤩
- [ ] No Linters error ✅ 
- [ ] All tests passed ⭐
- [ ] Amazing job working on the requested changes 💯

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**